### PR TITLE
signal: migrate fn is_sender_allowed(&self, sender: &str) -> bool → AllowlistAspect (archetype A)

### PR DESCRIPTION
**Pull request by yijunyu**
## Summary

- **Base branch:** `master`
- **What changed and why:** This branch is the chain tip of the 24-PR migration that replaces hand-rolled `is_*_allowed` predicates with the shared `aspect_std::AllowlistAspect`. Because PRs are stacked on each other (#6638 → #6778 → … → #6799 → #6800), the diff against `master` shows **all 16 channels' cumulative migration**, not just `signal`. The reviewer was correct to flag the scope mismatch — see "Cumulative scope" below.
- **Scope boundary:** This PR does NOT change identity resolution (e.g. WhatsApp LID→phone, Telegram `@`-prefix stripping happens before the allowlist check), webhook/SSE parsing, downstream message routing, LLM dispatch, or response sending. The aspect owns *only* the matching half of allowlist enforcement.
- **Blast radius:** All 16 messaging channels with an allowlist (Discord, Telegram, Slack, Matrix, Email, Gmail Push, WhatsApp, WhatsApp Web, IRC, Lark, Nostr, Mattermost, Mochat, Nextcloud Talk, QQ, Twitter, Signal, iMessage, dingtalk, wecom, wati, linq, wechat, discord_history). If `AllowlistAspect::is_allowed` regresses, every channel breaks at once. Mitigated by (a) per-channel parity tests against the old predicate and (b) the aspect-std test suite that travels with the crate.
- **Linked issue(s):** none
- **Labels:** `risk: high`, `size: L`, `type: refactor`, `area: channels`

## Cumulative scope — full 16-channel migration

This PR is **chain tip #21 of 24**. Each previous PR in the stack migrates one channel; the cumulative diff against `master` therefore contains every prior channel. Per-channel archetype + safety rationale:

| Channel | Archetype | Allowlist semantics preserved by |
|---|:-:|---|
| discord (#6638) | A | flat-list + `*` byte-equality |
| telegram (#6778) | C | `@`-prefix stripping + trim normalization at construction and check |
| wechat (#6779) | A | flat-list + `*` byte-equality |
| email (#6780) | C | shared `email_allowlist_match` predicate (domain `@example.com`, bare `example.com`, full email) |
| gmail-push (#6781) | C | same `email_allowlist_match` predicate as `email` |
| whatsapp-web (#6782) | C | `whatsapp_phone_match` two-way phone normalization |
| irc (#6783) | B | `with_normalizer(\|s\| s.to_ascii_lowercase())` |
| nostr (#6784) | D→A | bech32 npub canonicalized to hex at constructor; reduces to archetype A |
| matrix (#6785) | B | `with_normalizer(\|s\| s.to_ascii_lowercase())` |
| mattermost (#6786) | A | flat-list + `*` byte-equality |
| twitter (#6787) | A | flat-list + `*` byte-equality |
| wecom (#6788) | A | flat-list + `*` byte-equality |
| dingtalk (#6789) | A | flat-list + `*` byte-equality |
| qq (#6790) | A | flat-list + `*` byte-equality |
| slack (#6791) | A | flat-list + `*` byte-equality |
| lark (#6792) | A | flat-list + `*` byte-equality |
| nextcloud-talk (#6793) | A | flat-list + `*` byte-equality |
| mochat (#6794) | A | flat-list + `*` byte-equality |
| wati (#6795) | A | flat-list + `*` byte-equality |
| linq (#6796) | A | flat-list + `*` byte-equality |
| whatsapp (#6797) | A | flat-list + `*` byte-equality |
| discord-history (#6798) | presence-gate→A | empty-list = allow-all preserved via constructor-side `empty → ["*"]` canonicalization |
| **signal (this PR)** | **A** | **flat-list + `*` byte-equality** |
| imessage (#6800) | B | `with_normalizer(\|s\| s.to_ascii_lowercase())` |

The signal channel itself is archetype A: a flat `Vec<String>` of allowed sender IDs with `*` as wildcard, byte-equality otherwise. No normalization required; production `allowed_from` entries are E.164 phone numbers consumed verbatim.

## What changed (this channel)

- Replaced `allowed_from: Vec<String>` field with `allowlist: AllowlistAspect`.
- Deleted the 6-LOC `is_sender_allowed` predicate.
- Production caller (`SignalChannel::listen` envelope path) routes through `self.allowlist.is_allowed(sender)`.
- Unit tests updated to assert against `allowlist.is_allowed("…")` instead of field inspection.

## Supply-chain question (resolution)

The reviewer's blocker on sourcing `aspect-std` from `github.com/yijunyu/aspect-rs` is resolved by the latest commit on this branch (and all 23 other chain branches): **`aspect-std = "0.1.1"` is now on crates.io** (published 2026-05-24), so the dep is registry-sourced, immutable, and visible to `cargo audit` / `cargo deny`. The git-pinned source has been removed.

- `aspect-core v0.1.1` — https://crates.io/crates/aspect-core/0.1.1
- `aspect-runtime v0.1.1` — https://crates.io/crates/aspect-runtime/0.1.1
- `aspect-macros v0.1.1` — https://crates.io/crates/aspect-macros/0.1.1
- `aspect-std v0.1.1` — https://crates.io/crates/aspect-std/0.1.1 (contains `AllowlistAspect`)

The pre-aspect predicate was 6 LOC per channel × 25 sites = 147 LOC of duplicated control-flow. The aspect adds 4 new crates to the dep graph (~120 LOC of `AllowlistAspect` itself, ~600 LOC of `aspect-core` + `aspect-runtime` + `aspect-macros` foundation). The trade-off is centralized test coverage + uniform fix shape (a `with_normalizer`/`with_matcher` plug-point at one place) vs. transitive surface. We have measured the cost (M5: +5,824 bytes stripped binary across all 16 channels = +222 B amortized per migrated channel; build time inside noise).

## Validation Evidence

Local validation (commands from the test plan, tail output verbatim):

```
$ cargo build -p zeroclaw-channels --features channel-signal
   Compiling zeroclaw-channels v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-channels)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 23.41s

$ cargo test -p zeroclaw-channels --features channel-signal --lib signal::
test signal::tests::specific_sender_allowed ... ok
test signal::tests::strips_trailing_slash ... ok
test signal::tests::sse_envelope_deserializes_group ... ok
test signal::tests::sse_envelope_deserializes ... ok
test signal::tests::unknown_sender_denied ... ok
test signal::tests::wildcard_allows_anyone ... ok
test signal::tests::process_envelope_skips_attachment_only ... ok
test result: ok. 35 passed; 0 failed; 0 ignored; 0 measured; 956 filtered out; finished in 0.00s

$ cargo clippy -p zeroclaw-channels --features channel-signal --lib --no-deps -- -D warnings
    Checking aspect-core v0.1.1
    Checking aspect-std v0.1.1
    Checking zeroclaw-channels v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-channels)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 30.81s
```

- **Commands run and tail output:** the three above; all pass with zero warnings.
- **Beyond CI — what I manually verified:** the signal-specific `wildcard_allows_anyone` / `specific_sender_allowed` / `unknown_sender_denied` tests assert the parity contract directly. I did NOT replay against a live Signal CLI bridge — no isolated test instance available.
- **Skipped:** `cargo fmt --check` and `cargo test --workspace` are not in the test plan because the relevant invariants are channel-local; the workspace-level checks belong to the merge gate.

## Security & Privacy Impact

- New permissions, capabilities, or file system access scope? **No.**
- New external network calls? **No.**
- Secrets / tokens / credentials handling changed? **No.**
- PII, real identities, or personal data in diff, tests, fixtures, or docs? **No** — test fixtures use synthetic phone numbers (`+1111111111`).

This is an access-control refactor: it changes *where* the allowlist check is implemented but not *what* it enforces. Per-channel parity tests assert the new path produces the same allow/deny verdict as the old predicate across every truth-table case (wildcard, empty list, specific match, unknown, case sensitivity, substring rejection).

## Compatibility

- Backward compatible? **Yes** at the channel config surface — the `allowed_from` / `allowed_users` / `allowed_numbers` config fields are still consumed; their values are passed to `AllowlistAspect::new(...)` at channel construction.
- Config / env / CLI surface changed? **No.** Existing operator config files continue to work unchanged.
- Upgrade steps for existing users: none required.

## Rollback (risk: high)

- **Fast rollback command/path:** `git revert <merge-sha>` reverts the entire 24-PR chain as one merge commit (this is the chain tip; the PRs are designed to be reverted as a unit, not piecewise — partial revert would leave the workspace in a state where some channels reference `AllowlistAspect` and the dep is gone).
- **Feature flags or config toggles:** None. The aspect is statically linked into every channel that's enabled.
- **Observable failure symptoms:**
  - Allowlist-deny false positives: legitimate users denied at inbound. Grep logs for `is_allowed → false` followed by message drop, or look for `unauthorized sender` style messages in the channel-level traces.
  - Allowlist-bypass false negatives: unauthorized users let through. Watch for unexpected senders appearing in `chat_history` rows of channels that had non-wildcard allowlists.
  - In either case the metric to watch is per-channel message-drop rate (delta from steady-state in the hour after deploy).

## Note on the clone warning

The reviewer correctly flagged that `AllowlistAspect::clone` zeros out the `normalizer` and `matcher` fields. This is a known landmine for future use but does not affect any current channel: the only channel that clones (`discord::with_allowlist()` for the reaction-worker secondary constructor) uses plain `AllowlistAspect::new()` with no normalizer or matcher. We have filed this as a follow-up against `aspect-std` to either (a) make `Clone` carry the closures (requires `dyn Fn` to be `Send + Sync + Clone`-able via `Arc`), or (b) emit a compile-time warning when `clone()` is called on an aspect with non-`None` closures. Until then the safe pattern is to share the parent `Arc` (as `with_allowlist` already does for discord) rather than clone the aspect.

### CODE REVIEW

**Review by theonlyhennygod — CHANGES_REQUESTED** [review]
Reviewing PR #6799 — `signal: migrate fn is_sender_allowed → AllowlistAspect (archetype A)`.

I've read the full diff (25 files, +881/-500), the upstream `aspect-std::AllowlistAspect` source at the pinned rev, the PR body and test plan, the `AGENTS.md` risk tiers and single-source-of-truth rule, and the foundations documents. This is a high-risk access-control migration so I'm reviewing accordingly.

---

### 🔴 Blocking — scope mismatch between title/body and the actual diff

The PR title says "signal" and the body describes migrating "the `signal` channel's hand-rolled allowlist predicate." The test plan only exercises `--features channel-signal`. But the diff migrates **every channel** — discord, telegram, slack, matrix, email, whatsapp, irc, lark, nostr, and 13 others — across 25 files.

This is an access-control change touching every messaging integration in the codebase, and the PR body doesn't explain any of the non-signal migration decisions. A reviewer coming to this PR from the title expects ~1–2 files, not 25.

The PR body needs to accurately describe the full scope: which channels are migrated, what archetype each one uses (A vs. custom matcher vs. normalizer), and why the migration is safe for each channel's specific allowlist semantics (e.g., discord_history's empty-list = allow-all, email's domain matching, nostr's hex canonicalization, IRC's case-insensitivity, whatsapp_web's phone normalization).

---

### 🔴 Blocking — external dependency from PR author's personal repo

The dependency `aspect-std` is sourced from `https://github.com/yijunyu/aspect-rs` at a pinned commit. This raises three concrete problems:

1. **Supply-chain fragility.** A personal repo can be force-pushed, renamed, or deleted at any time. Even with a pinned `rev`, the ref is mutable. If the repo is restructured or the commit is garbage-collected, all ZeroClaw builds break. Published crates on crates.io are immutable once uploaded.

2. **No registry auditing.** Dependencies from git sources bypass the `cargo audit` / `cargo deny` supply-chain checks that the project relies on. There is no checksum verification from a trusted registry.

3. **Transitive weight.** `aspect-std` depends on `aspect-core`, which brings in the full AOP machinery (`Aspect` trait, `JoinPoint`, `ProceedingJoinPoint`, `AspectError`). ZeroClaw's channel allowlist check is a linear scan through a `Vec<String>` — the AOP framework adds complexity and transitive dependency surface for a problem that a standalone `struct Allowlist { entries: Vec<String>, ... }` in `zeroclaw-infra` or `zeroclaw-channels` would solve without external dependencies.

Per AGENTS.md: "Do not add heavy dependencies for minor convenience." The `AllowlistAspect` pattern is sound for deduplication, but it doesn't need to come from an external AOP framework.

**Resolution path:** Either (a) vendor or inline the `AllowlistAspect` struct (it's ~120 LOC including tests, no external dependencies beyond `parking_lot` which is already in-tree), or (b) publish `aspect-std` to crates.io so it's auditable and immutable, then re-introduce it as a proper registry dependency.

---

### 🔴 Blocking — PR body missing required template sections

The PR uses a non-standard body format. The following required sections from `.github/pull_request_template.md` are missing or incomplete:

- **Validation Evidence:** The test plan lists commands but shows no output. Per the template: "paste literal output (tails, failures, warnings — not 'all passed')."
- **Security & Privacy Impact:** Not present. This is a high-risk PR replacing access-control logic across 20+ channels — the section is mandatory.
- **Compatibility:** Not present. The config surface changes (the `allowed_users`/`allowed_from`/`allowed_numbers` fields are consumed differently).
- **Rollback:** Not present. Required for `risk: medium` and above; this PR carries `risk: high`.
- **Scope boundary / Blast radius:** Not present. The PR body doesn't state what it explicitly does NOT change, making it impossible to assess blast radius.

---

### 🟡 Warning — `AllowlistAspect::clone` silently drops matcher and normalizer

The upstream `Clone for AllowlistInner` sets `normalizer: None` and `matcher: None` on clone. The Discord channel's `with_allowlist()` path clones the parent's `AllowlistAspect` for the reaction worker. This is safe today because Discord uses plain `AllowlistAspect::new()` without a normalizer or matcher.

But it's a landmine: if any future channel adds `with_normalizer(...)` or `with_matcher(...)` and then clones (e.g., for a worker or a test helper), the clone will silently lose the normalization/matching logic and fall back to byte-equal comparison. This would be an access-control bypass — users who should be allowed would be silently denied (fail-closed, at least, not fail-open).

The Discord `with_allowlist` doc comment should note this constraint explicitly, or the upstream should support cloning with the matcher preserved.

---

### 🟡 Warning — over-engineering relative to the problem being solved

The actual allowlist check in every channel is: "is this identity in the list, or does the list contain `*`?" This is a `Vec::iter().any()` call. The `AllowlistAspect` wraps this in an `Arc<RwLock<Vec<String>>>` with optional closures, implements the `Aspect` AOP trait, and depends on `aspect-core`'s `ProceedingJoinPoint` machinery.

For the current use case, a simple struct in `zeroclaw-infra` would provide the same deduplication benefit without the external dependency, the AOP framework coupling, or the `RwLock` overhead (most channels never mutate the allowlist after construction).

This isn't blocking — the abstraction works — but it's worth being deliberate about whether the AOP coupling earns its weight.

---

### 🟢 What looks good — parity tests

Every migrated channel has a comprehensive parity test that runs the new `AllowlistAspect.is_allowed()` against the old hand-rolled predicate for a full truth table covering wildcard, empty-list, case-sensitivity, substring rejection, and edge cases. The assertion messages include the inputs for easy debugging.

This is exactly the right way to validate a behavior-preserving access-control migration. If any future change to `AllowlistAspect` breaks the semantics, these tests will catch it immediately.

---

### 🟢 What looks good — per-channel semantic preservation

Each channel's specific allowlist semantics are correctly preserved:
- **Discord history**: empty list = allow-all is preserved via canonicalization to `["*"]`
- **Email/Gmail**: domain matching (`@example.com`, bare `example.com`, full email) extracted into a shared `email_allowlist_match` predicate used by both channels
- **IRC/Matrix**: case-insensitive comparison via `with_normalizer(|s| s.to_ascii_lowercase())`
- **Nostr**: public key canonicalization from bech32 npub to hex before storing in the aspect
- **WhatsApp Web**: phone normalization via a custom `whatsapp_phone_match` two-way predicate
- **Telegram**: `@`-prefix stripping and trim normalization at both construction and check time

The migration is faithful to each channel's original semantics.

---

### 🟢 What looks good — Discord reaction-worker shared state

The Discord `with_allowlist()` secondary constructor shares the parent's `Arc<AllowlistInner>` with the reaction-worker clone. This avoids a duplicate `Vec` allocation, is cheap to clone, and ensures runtime `add()` calls propagate to the worker. Clean solution to a real problem.

---

### Verdict

Requesting changes. The three blockers are: (1) the scope mismatch between title/body and the 25-file diff makes this impossible to review properly as described, (2) the external dependency from a personal git repo is a supply-chain risk that violates the project's dependency philosophy, and (3) the missing required PR template sections leave security, compatibility, and rollback unaddressed for a high-risk access-control PR.

The per-channel migrations themselves are technically sound — the parity tests and semantic preservation are excellent. I'd like to see this land, but with the scope accurately described, the dependency sourced safely, and the PR body complete.

**Comment by theonlyhennygod** [comment]
@yijunyu — left a detailed review with three blockers. The short version:

1. **Scope mismatch** — title/body say signal-only, but the diff migrates all 20+ channels. The body needs to describe the full scope and per-channel design decisions.
2. **External dep** — `aspect-std` from your personal git repo is a supply-chain risk. Either vendor/inline the \~120 LOC `AllowlistAspect` struct, or publish to crates.io first.
3. **Missing template sections** — Validation Evidence (with actual output), Security & Privacy Impact, Compatibility, and Rollback are all required for a `risk: high` access-control PR.

The per-channel migrations and parity tests are solid — happy to re-review once these are addressed.

File: crates/zeroclaw-channels/src/discord_history.rs [hunk 1]
```diff
@@ -15,7 +16,7 @@ use zeroclaw_memory::{Memory, MemoryCategory};
 pub struct DiscordHistoryChannel {
     bot_token: String,
     guild_id: Option<String>,
-    allowed_users: Vec<String>,
+    allowlist: AllowlistAspect,
```
Comment (singlerider) [inline crates/zeroclaw-channels/src/discord_history.rs:19]: DiscordHistory is no longer a supported channel. Is this file in the v0.8.0 beta code? I removed it.

**Review by singlerider — CHANGES_REQUESTED** [review]

**Comment by yijunyu** [comment]
@theonlyhennygod thanks for the detailed review. Addressing all three blockers + the clone warning:

**Blocker 1 (scope mismatch) — fixed.** You're right that the title/body said "signal" but the diff was the full chain. The PR body is now rewritten with the cumulative scope (table of 24 channels + per-channel archetype + safety rationale), the per-channel breakdown of why each migration is semantics-preserving, and an explicit scope-boundary statement. This PR is chain tip #21 of 24; the diff against `master` is intentionally cumulative because the PRs are stacked on each other (and intended to merge as one chain).

**Blocker 2 (external dep from personal git repo) — fixed.** As of 2026-05-24, `aspect-std v0.1.1` is published to crates.io and contains `AllowlistAspect`:

- https://crates.io/crates/aspect-core/0.1.1
- https://crates.io/crates/aspect-runtime/0.1.1
- https://crates.io/crates/aspect-macros/0.1.1
- https://crates.io/crates/aspect-std/0.1.1

The latest commit on this branch (and all 23 other chain branches) replaces the `git = "https://github.com/yijunyu/aspect-rs", rev = "…"` line with `aspect-std = "0.1.1"`. The dep is now registry-sourced, immutable, and visible to `cargo audit` / `cargo deny`. The personal-repo supply-chain concern is resolved.

On the "vendor inline" alternative you mentioned: I considered it. The reason I picked publish-to-crates.io instead is that this is the M1 evaluation for an ICSE 2027 paper on aspect-rs — vendoring would dissolve the dependency relationship being measured. With v0.1.1 on the registry, the dep graph is real but auditable, which matches what the paper claims to measure.

**Blocker 3 (missing template sections) — fixed.** The new PR body includes Validation Evidence (with verbatim cargo command tails, no "all passed" hand-waving), Security & Privacy Impact, Compatibility, Rollback, Scope boundary, and Blast radius. The risk level is `risk: high` and the rollback section addresses fast-rollback path, feature flags, and observable failure symptoms.

**Warning on `AllowlistAspect::clone` — acknowledged.** You're correct that `Clone for AllowlistInner` zeros the closures. No current channel hits this — the only `with_allowlist()` clone site (discord's reaction-worker constructor) uses plain `AllowlistAspect::new()` without normalizer/matcher, and the migration's clone semantics are documented as "share via Arc, do not clone the aspect itself if it carries closures." Filed as a follow-up against `aspect-std` to either (a) make `Clone` carry `Arc<dyn Fn …>` closures or (b) emit a compile-time warning on closure loss. Will land in `aspect-std v0.2.0`.

**Warning on over-engineering relative to the problem — acknowledged.** The M5 evaluation measures exactly this question: across the 16-channel migration the aspect adds +5,824 bytes to the stripped binary (+222 B per migrated channel, amortized) and no measurable wall-clock build cost. The trade-off is the M1/M2 finding: 25 hand-rolled sites collapse to one shared implementation with concentrated test density (the `any → all` mutation that silently shipped on 6 of 10 baseline variant-A channels is caught one-shot by the aspect's 13-case parity table). The full M1-M5 evaluation is in `aspect-rs-priv/docs/icse2027-data/` (private until paper submission); happy to share specific numbers if useful.

Ready for re-review.

**Comment by yijunyu** [comment]
Closing. This channel already routes its allowlist check through the in-tree `crate::allowlist::is_user_allowed(&peers, id, Match::…)` that landed in #6398 — the same consolidation this PR's `AllowlistAspect` migration was proposing, done in-tree with no external dependency and resolving the peers fresh from config per call (single source of truth).

Re-applying the aspect on top of that would replace working in-tree code with an external crate for no behavioral gain, so this PR is superseded.

The aspect still lands where `crate::allowlist`'s flat-list matching structurally can't reach — the channels whose identity matching needs normalization the `Match` enum can't express. Those migrations are open as #7017 (whatsapp_web, E.164 phone) and #7018 (email + gmail_push, domain-class), using a stateless `aspect_std::is_allowed_in` (crates.io `aspect-std 0.1.2`) that holds no state and so honors the single-source-of-truth rule.

Thanks to the reviewers who looked at this stack.
